### ISSUE TITLE
[Superfences] Pass fenced code block config (hl_lines, linenums, ...) to custom formatter func

### ISSUE THREAD

**Issue by dbader** [facelessuser/pymdown-extensions#349]
Hi @facelessuser, I absolutely love your MD extensions. They're powering large parts of realpython.com and your documentation has been a joy to work with 😃 

I'm using the custom formatter functions in `SuperFences` to customize rendering and syntax highlighting. I'm looking to find a way to access fenced code block config like `hl_lines` and `linenums` from the formatter function.

As best as I can tell, a small change in the custom formatter function API to start passing through those config values would achieve this quite easily:

https://github.com/facelessuser/pymdown-extensions/blob/881e852ecfd06449fe608932c3abd4deffe2005e/pymdownx/superfences.py#L370

```python
    def process_nested_block(self, ws, content, start, end):
        """Process the contents of the nested block."""

        self.last = ws + self.normalize_ws(content)
        code = None
        for entry in reversed(self.extension.superfences):
            if entry["test"](self.lang):
                #  
                # TODO: Pass extra data like self.hl_lines to entry['formatter']:
                # 
                code = entry["formatter"](self.rebuild_block(self.code), self.lang)
                break
```

Do you think this is the best way to go about it or should I look elsewhere? Happy to put together a PR if you think the approach I suggested makes sense.

Thanks again!

— Dan

**Comment by facelessuser** [comment]
Hmm. I guess I always looked at the custom formatter as something for things like UML or other non-code like things.  In those cases, `hl_lines` and `linenums` have no meaning.  And since you can't really define custom options, it didn't make since to pass such things...

I'm going to make a guess, is this to provide a custom format for a JavaScript highlighter?

**Comment by dbader** [comment]
> I'm going to make a guess, is this to provide a custom format for a JavaScript highlighter?

Close! I'm doing some transforms on the source code inside the fenced block, like removing / adding `>>>` interpreter prompts and then pass it to `pymdownx.highlight.Highlight.highlight()` to render the syntax highlighted HTML.

As I'm writing this I realized I could put the config data *into* the code block and then parse it out in the custom formatter:

    ```custom_lang
    #$ hl_lines="1 2 3"
    lorem ipsum dolor
    sit amet ...
    ```

Thanks for the rubber ducking session lol... It would still be nice to standardize the syntax and have access to the config info in the custom formatter func, but I'd be equally happy with a workaround.

**Comment by facelessuser** [comment]
If I was approaching this, there are a couple of things I would take into consideration.

- Most custom formatters (currently) don't need the highlight options, so I would make it...well optional.  I imagine extended `custom_fences` dictionary to have a key named something like `pass_hl_options`.  Probably access it with `.get` so you can assign a default value of `False` if the key isn't present.  This will help preserve backwards compatibility.

- Then you would conditionally pass the options into the format only if we originally got `True` for `pass_hl_options` on **that** custom fence.  I don't want to force people to change existing formatters to accept additional option(s) they don't need, so it would be better to pass them only when required.

- When passing the options, I'd pass them as single parameter that is a dictionary.  Each option would be accessed via their normal key name.  I'd go ahead and pass all the available options just for the sake of consistency.

So in short, I'm not against such a feature. SuperFences essentially evolved from my own needs at the time, but if we can extend it in a generalized way that can assist additional, advanced use cases, I'm fine with that.  Since it could help with specific JavaScript highlighter options, I think it makes sense to add something like this.

**Comment by facelessuser** [comment]
> As I'm writing this I realized I could put the config data into the code block and then parse it out in the custom formatter:

This is essentially what CodeHilite, Python Markdown's builtin highlighter, does.  I'm not a big fan of this as any syntax you pick may conflict with some language.  With that said, it may be a reasonable workaround, but I am more than happy to take a pull request for something like I described above.

When I have time, I don't mind throwing it together, but  a pull request will get it in much faster 😉 .

**Comment by dbader** [comment]
> If I was approaching this, there are a couple of things I would take into consideration.

Cheers, you're right, we'd definitely want to introduce this in a way that retains backwards compatibility with existing custom formatters 😃 

I like your idea of passing a dict to the formatter func that can hold any number of options that might be introduced in the future.

On that note, would it make sense to make this a generic config system from the get go that parses a simple `key=value` syntax and forwards them to the formatter?

For example:

    ```lang hl_lines="1 3" my_key="yes" linenums="1"
    lorem ipsum
    dolor sit amet
    ```

→ This would result in the following config dict passed to the formatter:

```python
{'hl_lines': '1 3', 'my_key': 'yes', 'linenums': '1'}
```

**Comment by facelessuser** [comment]
Currently, it doesn't just look for keys, it looks for keys and values that adhere to a specific format.  To do a generalized list, you have to relax the format.  Poorly formatted values and erroneous keys will then be allowed unless you provide a validation hook for a given custom language.  Ideally if the hook fails in validating the keys and values, it should prevent process the block.  To go a generalized route, I would most likely require a such a validation hook as I want it to skip blocks that are malformed.

Doing this increases the work needed and adds a bit more complexity.  There would most likely be some refactoring required to make this happen.

**Comment by facelessuser** [comment]
But also, there are some options that are globally accepted like the newer `tab` key, and there may be more in the future (probably not, but the possibility is open)...I'm open to a proposal. I haven't really thought much about all this yet 😄 .  While there is the ability for custom blocks, its design is still fairly code biased.

**Comment by facelessuser** [comment]
Thinking a bit more, I imagine you could generally parse all the keys into a dictionary, extract globals first (validate and remove from the dictionary), then pass the remaining language specific options to the per language validate hook.  Anyways, that's enough brain storming from me today 🙂 .

**Comment by facelessuser** [comment]
Don't worry about a pull.  I'm knee deep in the changes now.

**Comment by dbader** [comment]
You're amazing @facelessuser, thank you! This helps us out a ton. Are you on Twitter? I'd love to give you a shoutout from the Real Python account and tell people about your amazing project.

**Comment by facelessuser** [comment]
Thanks :smile:. It's a work in progress, and all I've tested currently is that I didn't break anything. So there is likely to be some refactoring, testing, and bug fixing to come.

As far as Twitter is concerned, I'm there, but under something not normally associated with my known handles. Feel free to give a shout out via a link to the project, I'm okay with it not directly linked to my Twitter handle. I kind of like the separation.

**Comment by facelessuser** [comment]
@dbader, pull #350 will be your opportunity to give feedback into the direction this feature is taking.  Your opinions are welcome during the development of it. In general, I think the direction I *want* to take is currently represented and appears to be passing basic sanity checks.  This evening I'll probably test actual custom options and verify that they they make it to custom formatters.

**Comment by dbader** [comment]
Cheers @facelessuser I left some comments on #350 :)
